launch_testing/test/launch_testing/test_proc_info_assertions.py

```diff
@@ -0,0 +1,137 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+import signal
+import sys
+import types
+import unittest
+
+import ament_index_python
+import launch
+import launch.actions
+import launch.events.process
+from launch_testing.loader import LoadTestsFromPythonModule
+from launch_testing.test_runner import LaunchTestRunner
+import launch_testing.util
+
+
+def test_wait_for_shutdown():
+
+    def generate_test_description(ready_fn):
+        TEST_PROC_PATH = os.path.join(
+            ament_index_python.get_package_prefix('launch_testing'),
+            'lib/launch_testing',
+            'good_proc'
+        )
+
+        good_process = launch.actions.ExecuteProcess(
+                cmd=[sys.executable, TEST_PROC_PATH],
+        )
+
+        # Let 'good_process' run for 10 seconds, then terminate it
+        return launch.LaunchDescription([
+            good_process,
+            launch_testing.util.KeepAliveProc(),
+            launch.actions.TimerAction(
+                period=10.0,
+                actions=[
+                    launch.actions.EmitEvent(
+                        event=launch.events.process.SignalProcess(
+                            signal_number=signal.SIGINT,
+                            process_matcher=lambda proc: proc is good_process
+                        ) 
+                    )
+                ]
+            ), 
+            launch.actions.OpaqueFunction(function=lambda context: ready_fn())
+        ]), {'good_process': good_process}
+
+    # This is kind of a weird test-within-a-test, but it's the easiest way to get
+    # all of the proc_info handlers hooked up correctly without digging deep into
+    # the guts of the runner
+    class ProcInfoTests(unittest.TestCase):
+
+        def test_01_check_running_process(self, proc_info, good_process):
+            try:
+                proc_info.assertWaitForShutdown(good_process, timeout=3)  # Should raise
+                assert False, "assertWaitForShutdown did not raise"
+            except AssertionError as e:
+                assert "Timed out waiting for process" in str(e)
+
+        def test_02_check_when_process_exits(self, proc_info, good_process):
+            proc_info.assertWaitForShutdown(good_process, timeout=15)
+
+    test_module = types.ModuleType('test_module')
+    test_module.generate_test_description = generate_test_description
+    test_module.ProcInfoTests = ProcInfoTests
+
+    runner = LaunchTestRunner(
+        LoadTestsFromPythonModule(test_module)
```

Comment:
@pbaughman meta: can't we feed test cases in other, more direct way?

Reply:
@hidmic Are you thinking a factory method that the tests use, or are you thinking something else?  I agree this is getting a little out of hand.  The following test files do this or something similar:
  * test_proc_info_assertions.py
  * test_xml_output.py (doesn't actually run the tests though)
  * test_resolve_process.py
  * test_runner_results.py

I want to use the real 'LaunchTestRunner' from the tests, which needs a `TestRun` object which is somewhat tightly coupled to the `LoadTestsFromPythonModule` function that is normally used to create them.

I could create a different factory function that makes TestRun objects without so much boilerplate.  Would that be a test-only thing, or would it be worth making that part of the real API?  Again - I agree that this could be better - open to input

Reply:
> I could create a different factory function that makes TestRun objects without so much boilerplate.

That'd be ideal :)

Reply:
I think no matter what we do, we're going to have to stuff things into a module so we don't have to replicate the discovery logic in https://github.com/python/cpython/blob/master/Lib/unittest/loader.py#L98-L136

At the end of the day, the unittest loader expects a module.

I'm going to make this function available to the tests.  We can decide later if the function should be part of the regular API instead.  I worry that right now I don't understand the edge cases well enough to support this as part of the regular API.